File: programmers/javascript/level3/N으로표현.js
const N = 5;
const number = 12;

function solution(N, number) {
  const cache = new Array(9).fill(0).map((el) => new Set());
  console.log(cache);
  for (let i = 1; i < 9; i++) {
    cache[i].add("1".repeat(i) * N);
    for (let j = 1; j < i; j++) {
      for (const arg1 of cache[j]) {
        for (const arg2 of cache[i - j]) {
          cache[i].add(arg1 + arg2);
          cache[i].add(arg1 - arg2);
          cache[i].add(arg1 * arg2);
          cache[i].add((arg1 / arg2) >> 0);
        }
      }
    }
    console.log(cache);
    if (cache[i].has(number)) return i;
  }
  return -1;
}

console.log(solution(N, number));

File: python/programmers/level3/N으로 표현.py
def solution(N, number):
    dp = [0, {N}]

    if N == 1 or N == number: return N % number + 1

    for i in range(2, 9):
        setA = {int(str(N) * i)}
        for j in range(1, i // 2 + 1):
            for x in dp[j]:
                for y in dp[i - j]:
                    setA.add(x + y)
                    setA.add(x - y)
                    setA.add(y - x)
                    setA.add(x * y)
                    if x != 0:
                        setA.add(y // x)
                    if y != 0:
                        setA.add(x // y)
        if number in setA:
            return i
        dp.append(setA)
    return -1


print(solution(5, 12))
